src/Bicep.Core/Parsing/BaseParser.cs

```diff
@@ -384,6 +406,11 @@ private SyntaxBase FunctionCallOrVariableAccess(ExpressionFlags expressionFlags)
                 return new LambdaSyntax(new LocalVariableSyntax(identifier), arrow, expression);
             }
 
+            if (HasExpressionFlag(expressionFlags, ExpressionFlags.TypeExpression))
+            {
+                return LanguageConstants.DeclarationTypes.ContainsKey(identifierToken.Text) ? new SimpleTypeSyntax(identifierToken) : new TypeAccessSyntax(identifier);
```

Comment:
What's the reason to differentiate between the two? Couldn't we treat primitive types as pre-existing? (e.g. similar to the way built-in namespaces are handled)

Reply:
This approach basically treats primitive type names as language keywords but only within type expressions. It also had the upside of letting me more or less leave output serialization untouched, since the only constraint we allow on those in ARM is `type`.

I went back and forth on whether to put this check here or in the `NameBindingVisitor` (i.e., treating primitive types as pre-existing) and ended up going with this approach because it touched fewer lines of existing code. I can switch tactics if you feel strongly about this; I am not at all attached to having this fork in the parser.

Reply:
I'm fine keeping as-is for now. In the long run, it would be nice to try and simplify this - conceptually it's quite similar to built-in namespaces (although that's more complicated as we now allow aliasing).

Reply:
Tracking in https://github.com/Azure/bicep/issues/8811